usr.bin/mdo/mdo.c

```diff
@@ -19,52 +20,277 @@
+			print_rule = true;
+
+			for (int j = i; j < argc - 1; j++) {
+				argv[j] = argv[j+1];
+			}
+			argc--;
+			break;
+		}
+	}
+
+	while ((ch = getopt(argc, argv, "u:ig:G:s:U:R:E:P:Q:h")) != -1) {
 		switch (ch) {
 		case 'u':
 			username = optarg;
 			break;
 		case 'i':
 			uidonly = true;
 			break;
+		case 'g':
+			primary_group = optarg;
+			break;
+		case 'G':
+			supp_groups_str = optarg;
+			break;
+		case 's':
+			group_mod_str = optarg;
+			break;
+		case 'U':
+			ruid_str = optarg;
+			break;
+		case 'R':
+			svuid_str = optarg;
+			break;
+		case 'E':
+			euid_str = optarg;
+			break;
+		case 'P':
+			rgid_str = optarg;
+			break;
+		case 'Q':
+			svgid_str = optarg;
+			break;
+		case 'h':
+			usage();
 		default:
 			usage();
 		}
 	}
+
+	if (uidonly && (primary_group || supp_groups_str || group_mod_str))
+		errx(EXIT_FAILURE, "-i cannot be used with -g, -G, or -s");
+
 	argc -= optind;
 	argv += optind;
 
-	if ((pw = getpwnam(username)) == NULL) {
-		if (strspn(username, "0123456789") == strlen(username)) {
-			const char *errp = NULL;
-			uid_t uid = strtonum(username, 0, UID_MAX, &errp);
-			if (errp != NULL)
-				err(EXIT_FAILURE, "invalid user ID '%s'",
-				    username);
-			pw = getpwuid(uid);
+	if (username) {
+		if ((pw = getpwnam(username)) == NULL) {
+			if (strspn(username, "0123456789") == strlen(username)) {
+				const char *errp = NULL;
+				uid_t uid = strtonum(username, 0, UID_MAX, &errp);
+				if (errp != NULL)
+					err(EXIT_FAILURE, "invalid user ID '%s'", username);
+				pw = getpwuid(uid);
+			}
+			if (pw == NULL)
+				err(EXIT_FAILURE, "invalid username '%s'", username);
 		}
-		if (pw == NULL)
-			err(EXIT_FAILURE, "invalid username '%s'", username);
 	}
 
 	wcred.sc_uid = wcred.sc_ruid = wcred.sc_svuid = pw->pw_uid;
 	setcred_flags |= SETCREDF_UID | SETCREDF_RUID | SETCREDF_SVUID;
 
-	if (!uidonly) {
+	if (ruid_str) {
+		const char *errp = NULL;
+		wcred.sc_ruid = strtonum(ruid_str, 0, UID_MAX, &errp);
+		if (errp)
+			err(EXIT_FAILURE, "-U: invalid UID");
+		setcred_flags |= SETCREDF_RUID;
+	}
+	if (svuid_str) {
+		const char *errp = NULL;
+		wcred.sc_svuid = strtonum(svuid_str, 0, UID_MAX, &errp);
+		if (errp)
+			err(EXIT_FAILURE, "-R: invalid UID");
+		setcred_flags |= SETCREDF_SVUID;
+	}
+	if (euid_str) {
+		const char *errp = NULL;
+		wcred.sc_uid = strtonum(euid_str, 0, UID_MAX, &errp);
+		if (errp)
+			err(EXIT_FAILURE, "-E: invalid UID");
+		setcred_flags |= SETCREDF_UID;
+	}
+	if (rgid_str) {
+		const char *errp = NULL;
+		wcred.sc_rgid = strtonum(rgid_str, 0, GID_MAX, &errp);
+		if (errp)
+			err(EXIT_FAILURE, "-P: invalid GID");
+		setcred_flags |= SETCREDF_RGID;
+	}
+	if (svgid_str) {
+		const char *errp = NULL;
+		wcred.sc_svuid = strtonum(svgid_str, 0, GID_MAX, &errp);
+		if (errp)
+			err(EXIT_FAILURE, "-Q: invalid GID");
+		setcred_flags |= SETCREDF_SVGID;
+	}
+
+	if (primary_group) {
+		struct group *gr = getgrnam(primary_group);
+		if (gr)
```

Comment:
Style: `x` or `!x` as a condition is allowed only for booleans (whether variables that have `bool` as their types, or integer variables that are used as a boolean (old style)).  In particular, for pointers, these forms are not allowed, and only `p == NULL` or `p != NULL` should be used (thus, here, `gr != NULL`).

Not repeating it, but there are several occurrences of this below.